r/src/array.cpp

```diff
@@ -223,6 +287,42 @@ inline SEXP BooleanArray_to_Vector(const std::shared_ptr<arrow::Array>& array) {
   return vec;
 }
 
+inline SEXP StringArray_to_Vector(const std::shared_ptr<arrow::Array>& array) {
+  auto n = array->length();
+  Rcpp::CharacterVector res(n);
+
+  const auto& buffers = array->data()->buffers;
+
+  auto p_offset = reinterpret_cast<const int32_t*>(buffers[1]->data()) + array->offset();
+  auto p_data = reinterpret_cast<const char*>(buffers[2]->data()) + *p_offset;
```

Comment:
Is this only happening when length() == 0 ? Are there any examples (tests perhaps) of what to do then ?

Reply:
It really depends. What you want is:

```
auto p_data = GetValues<const char*>(buffers[2], *p_offset);
```

We already have such a function https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/util-internal.h#L34. This needs to have a null check (so that is `buffer[i]` is null then it returns nullptr rather than invoking `->data()`)

Whether this happens depends on the input data